packages/extension/src/view/devtools/components/cookies/index.tsx

```diff
@@ -25,18 +25,74 @@ import React from 'react';
 import { useCookieStore } from '../../stateProviders/syncCookieStore';
 import CookiesListing from './cookiesListing';
 import CookiesLanding from './cookiesLanding';
+import { Button, ProgressBar } from '../../../design-system/components';
 
 const Cookies = () => {
-  const { selectedFrame } = useCookieStore(({ state }) => ({
+  const {
+    allowedNumberOfTabs,
+    contextInvalidated,
+    isCurrentTabBeingListenedTo,
+    loading,
+    returningToSingleTab,
+    selectedFrame,
+    changeListeningToThisTab,
+  } = useCookieStore(({ state, actions }) => ({
+    allowedNumberOfTabs: state.allowedNumberOfTabs,
+    contextInvalidated: state.contextInvalidated,
+    isCurrentTabBeingListenedTo: state.isCurrentTabBeingListenedTo,
+    loading: state.loading,
+    returningToSingleTab: state.returningToSingleTab,
     selectedFrame: state.selectedFrame,
+    changeListeningToThisTab: actions.changeListeningToThisTab,
   }));
 
+  if (
+    loading ||
+    (loading &&
+      isCurrentTabBeingListenedTo &&
+      allowedNumberOfTabs &&
+      allowedNumberOfTabs === 'single')
+  ) {
+    return (
+      <div className="w-full h-screen flex items-center justify-center overflow-hidden bg-white dark:bg-raisin-black">
+        <ProgressBar additionalStyles="w-full" />
+      </div>
+    );
+  }
+
+  if (
+    (isCurrentTabBeingListenedTo &&
+      allowedNumberOfTabs &&
+      allowedNumberOfTabs === 'single') ||
+    (allowedNumberOfTabs && allowedNumberOfTabs === 'unlimited')
+  ) {
+    return (
+      <div
+        className={`h-full ${selectedFrame ? '' : 'flex items-center'}`}
+        data-testid="cookies-content"
+      >
+        {selectedFrame ? <CookiesListing /> : <CookiesLanding />}
+      </div>
+    );
+  }
+
   return (
-    <div
-      className={`h-full ${selectedFrame ? '' : 'flex items-center'}`}
-      data-testid="cookies-content"
-    >
-      {selectedFrame ? <CookiesListing /> : <CookiesLanding />}
+    <div className="w-full h-screen overflow-hidden bg-white dark:bg-raisin-black">
+      <div className="w-full h-full flex flex-col items-center justify-center">
+        {!returningToSingleTab && !contextInvalidated && (
+          <p className="dark:text-bright-gray text-chart-label text-base mb-5">
+            This tool works best with a single tab.
+          </p>
+        )}
+        {contextInvalidated ? (
```

Comment:
Just wanted to mention that we're working on multiple things in this pull request. It would be really helpful if we could have separate pull requests for each of these issues. But don't worry, it's fine for now.